### PR TITLE
Move our references over to Dev15

### PR DESCRIPTION
**Pull request by jaredpar**
Note this is a cross cutting change.  [Internal PR](https://github.com/dotnet/roslyn-internal/pull/1486)

**Customer scenario**

Roslyn is not fully building and testing against our Dev15 binaries.  The code base needs to be switched over to fully validate our testing and unlock other bug fixes / perf improvements.

**Workarounds, if any**

No work arounds. 

**Risk**

Low.  This doesn't change how our product runs on customer machines.  This just changes our infrastructure so we are validating against the binaries we are shipping against. 

**Performance impact**

Low. Again won't change what we run against, just what we compile against.   This mostly affects the set of binaries that we reference.  No new binaries were referenced, just the versions were updated. 

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

This gaps is well known however it could not be fixed until several other systems were switched over to Dev15.  In particular our internal integration testing scenarios needed to be switched.  Those were blocked for several months trying to adjust to setup changes related to Willow.  Those were switched over late last week and this change was made as soon as that completed.

### CODE REVIEW

**Comment by jaredpar** [comment]
retest this please

**Comment by jaredpar** [comment]
retest this please

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-infrastructure for code review

File: build/Targets/Imports.targets [hunk 1]
```diff
@@ -90,6 +90,11 @@
   </Choose>
 
   <Choose>
+    <When Condition="'$(TargetFrameworkIdentifier)' == '.NETPortable' AND '$(OS)' != 'Windows_NT'">
+      <PropertyGroup>
+        <MSBuildTargetsFilePath>$(MSBuildExtensionsPath32)\Extensions\Microsoft\Portable\$(TargetFrameworkVersion)\Microsoft.Portable.$(MSBuildTargetsLanguageName).targets</MSBuildTargetsFilePath>
+      </PropertyGroup>
+    </When>
```
Comment (tmeschter) [inline build/Targets/Imports.targets]: Why is the above necessary?

File: build/Targets/Settings.props [hunk 2]
```diff
@@ -224,7 +224,7 @@
   <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props"
           Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props') AND '$(MSBuildProjectExtension)' != '.vcxproj'" />
 
-  <Import Project="$(RoslynDiagnosticsPropsFilePath)" Condition="'$(UseRoslynAnalyzers)' == 'true'" />
+  <Import Project="$(RoslynDiagnosticsPropsFilePath)" Condition="'$(UseRoslynAnalyzers)' == 'true' AND Exists('$(RoslynDiagnosticsPropsFilePath)')" />
```
Comment (tmeschter) [inline build/Targets/Settings.props]: Under what circumstances will this not exist?

File: build/config/RepoUtilData.json [hunk 3]
```diff
@@ -1,26 +1,9 @@
...
-    "Microsoft.VisualStudio.ImageCatalog": "14.3.25407",
-    "Microsoft.VisualStudio.Shell.Immutable.10.0": "10.0.30319",
-    "Microsoft.VisualStudio.ComponentModelHost": "14.3.25407-alpha",
-    "Microsoft.VisualStudio.Text.UI.Wpf": "14.3.25407",
-    "Microsoft.VisualStudio.Text.Logic": "14.3.25407", 
-    "Microsoft.VisualStudio.Text.Data": "14.3.25407", 
-    "Microsoft.VisualStudio.Text.UI": "14.3.25407",
-    "Microsoft.VisualStudio.Utilities": "14.3.25407",
-    "Microsoft.VisualStudio.Validation": "14.1.111",
-    "Microsoft.VisualStudio.Threading": "14.1.111",
```
Comment (tmeschter) [inline build/config/RepoUtilData.json]: Why did these all go away, instead of being updated to version 15.x?

File: src/Compilers/Core/MSBuildTaskTests/IntegrationTests.cs [hunk 4]
```diff
@@ -26,7 +26,10 @@ public class IntegrationTests : TestBase
         static IntegrationTests()
         {
             s_msbuildDirectory = TestHelpers.GetMSBuildDirectory();
-            s_msbuildExecutable = Path.Combine(s_msbuildDirectory, "MSBuild.exe");
+            if (s_msbuildExecutable != null)
+            {
+                s_msbuildExecutable = Path.Combine(s_msbuildDirectory, "MSBuild.exe");
+            }
```
Comment (tmeschter) [inline src/Compilers/Core/MSBuildTaskTests/IntegrationTests.cs]: So if `s_msbuildExecutable` is `null` we leave it alone, but if it's not `null` we explicitly overwrite it? Why?

File: src/EditorFeatures/Test2/EditorServicesTest2.vbproj [hunk 5]
```diff
@@ -322,9 +322,15 @@
   </ItemGroup>
   <ItemGroup>
     <None Include="project.json" />
+    <None Include="..\..\Compilers\Core\MSBuildTask\Microsoft.CSharp.Core.targets">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Include="..\..\Compilers\Core\MSBuildTask\Microsoft.VisualBasic.Core.targets">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
```
Comment (tmeschter) [inline src/EditorFeatures/Test2/EditorServicesTest2.vbproj]: Is this to allow tests that utilize MSBuild to run correctly?

File: build/Defaults/Desktop/app.config [hunk 6]
```diff
@@ -73,6 +73,54 @@
         <assemblyIdentity name="Newtonsoft.Json" publicKeyToken="30ad4fe6b2a6aeed" culture="neutral"/>
         <bindingRedirect oldVersion="0.0.0.0-8.0.0.0" newVersion="8.0.0.0"/>
       </dependentAssembly>
+
+      <!-- Even though all of our references are to Dev15 binaries we are still pulling in the interactive window
```
Comment (tmat) [inline build/Defaults/Desktop/app.config]: Update to the latest IW package instead?

**Comment by jmarolf** [comment]
@jaredpar don't these changes need to be made on ETA also?

**Comment by jaredpar** [comment]
@jmarolf top of the PR I linked to the roslyn-internal PR.  Is there something more you were expecting?

File: src/EditorFeatures/Test2/EditorServicesTest2.vbproj (hunk 5 above)
Comment (jaredpar) [inline src/EditorFeatures/Test2/EditorServicesTest2.vbproj]: Correct.

File: build/Targets/Imports.targets (hunk 1 above)
Comment (jaredpar) [inline build/Targets/Imports.targets]: https://github.com/Microsoft/msbuild/issues/1533

Next change will get a link to that added into the targets file.

File: build/Targets/Settings.props (hunk 2 above)
Comment (jaredpar) [inline build/Targets/Settings.props]: If try to restore our projects on a 100% clean machine in the wrong order this won't exist.  This makes the error message in that case a bit clearer.  

I hit this problem on this change 😦

File: build/config/RepoUtilData.json (hunk 3 above)
Comment (jaredpar) [inline build/config/RepoUtilData.json]: Prior to this change there were two versions of all of these NuGet packages used in our repo:

- Dev14
- Dev15

This entry in the config file essentially told RepoUtil two things:

- This difference is intentional, don't count it as an error
- If you're asked to upgrade MS.VS.Text.Logic to a new version, do not upgrade anything that is marked as Dev14

Now that there is a single version, none of these entries are necessary.

File: build/Defaults/Desktop/app.config (hunk 6 above)
Comment (jaredpar) [inline build/Defaults/Desktop/app.config]: @tmat what is the latest version?

**Comment by jmarolf** [comment]
@jaredpar ah, missed that, nevermind

File: src/Compilers/Core/MSBuildTaskTests/IntegrationTests.cs (hunk 4 above)
Comment (mattwar) [inline src/Compilers/Core/MSBuildTaskTests/IntegrationTests.cs]: I suspect the check should read:  if (s_msbuildDirectory != null)

**Comment by mattwar** [comment]
Will this change cause the roslyn package that includes workspace.desktop (with msbuild workspace) to pull in a full msbuild with it?

**Comment by jaredpar** [comment]
@mattwar no.  This will only change our test libraries.

**Comment by jmarolf** [comment]
@jaredpar is this being considered for RC3?  https://github.com/dotnet/roslyn/issues/16040 requires this PR to work

File: src/Workspaces/Remote/Razor/project.json [hunk 7]
```diff
@@ -1,8 +1,8 @@
 ﻿{
   "dependencies": {
-    "StreamJsonRpc": "0.12.32-alpha-g90be50f449"
+    "StreamJsonRpc": "1.0.2-rc"
```
Comment (heejaechang) [inline src/Workspaces/Remote/Razor/project.json]: 👍

File: src/Workspaces/Remote/ServiceHub/project.json [hunk 8]
```diff
@@ -1,8 +1,9 @@
 ﻿{
   "dependencies": {
-    "StreamJsonRpc": "0.12.32-alpha-g90be50f449"
+    "StreamJsonRpc": "1.0.2-rc",
```
Comment (heejaechang) [inline src/Workspaces/Remote/ServiceHub/project.json]: 👍

**Comment by jaredpar** [comment]
@jmarolf it's being considered for RTW.  If you think that other bug has priority to drive this to RC3 I'm happy to discuss it and loop you into the thread.

**Comment by jaredpar** [comment]
CC @MattGertz for ask mode approval

File: build/Defaults/Desktop/app.config (hunk 6 above)
Comment (tannergooding) [inline build/Defaults/Desktop/app.config]: Looks to be 2.0.0-rc3-61304-01: https://dotnet.myget.org/gallery/interactive-window

File: build/config/RepoUtilData.json [hunk 9]
```diff
@@ -1,26 +1,9 @@
 ﻿{
   "fixedPackages": {
-    "Microsoft.VisualStudio.Language.Intellisense": "14.3.25407",
     "Microsoft.VSSDK.BuildTools": [ "14.3.25420", "15.0.25929-RC2" ],
```
Comment (tannergooding) [inline build/config/RepoUtilData.json]: Where are we still using the older BuildTools (and why)?

File: build/config/RepoUtilData.json [hunk 10]
```diff
@@ -1,26 +1,9 @@
...
     "Newtonsoft.Json": "9.0.1",
     "System.Reflection.Metadata": "1.0.21",
     "System.Collections.Immutable": "1.1.36",
```
Comment (tannergooding) [inline build/config/RepoUtilData.json]: Where are we using this legacy version of S.C.Immutable?

File: build/config/RepoUtilData.json [hunk 11]
```diff
@@ -1,26 +1,9 @@
 ﻿{
   "fixedPackages": {
-    "Microsoft.VisualStudio.Language.Intellisense": "14.3.25407",
     "Microsoft.VSSDK.BuildTools": [ "14.3.25420", "15.0.25929-RC2" ],
     "Newtonsoft.Json": "9.0.1",
     "System.Reflection.Metadata": "1.0.21",
```
Comment (tannergooding) [inline build/config/RepoUtilData.json]: Where are we using this legacy version of S.R.Metadata?

File: src/Compilers/Extension/project.json [hunk 12]
```diff
@@ -1,8 +1,12 @@
 ﻿{
   "dependencies": {
     "EnvDTE": "8.0.0",
-    "Microsoft.VisualStudio.Shell.14.0": "14.3.25407",
-    "Microsoft.VisualStudio.Shell.Interop.10.0": "10.0.30319"
+    "Microsoft.VisualStudio.Shell.15.0": "15.0.26014-alpha",
```
Comment (tannergooding) [inline src/Compilers/Extension/project.json]: There is an 'official' RC3 package we should probably be using instead: https://www.nuget.org/packages/Microsoft.VisualStudio.Shell.15.0/15.0.26109-RC3

File: src/Compilers/Extension/project.json [hunk 13]
```diff
@@ -1,8 +1,12 @@
...
+    "Microsoft.VisualStudio.Shell.Interop.8.0": "8.0.50727",
+    "Microsoft.VisualStudio.Shell.Interop.9.0":  "9.0.30729",
+    "Microsoft.VisualStudio.Shell.Interop.10.0": "10.0.30319",
+    "Microsoft.VisualStudio.Shell.Framework": "15.0.26014-alpha",
```
Comment (tannergooding) [inline src/Compilers/Extension/project.json]: Same goes here (https://www.nuget.org/packages/Microsoft.VisualStudio.Shell.Framework/15.0.26109-RC3) and the rest of our `-alpha` entries below.

File: src/EditorFeatures/Core/project.json [hunk 14]
```diff
@@ -1,19 +1,19 @@
...
-    "Microsoft.VisualStudio.Language.Intellisense": "14.3.25407",
-    "Microsoft.VisualStudio.Imaging": "14.3.25407",
+    "Microsoft.VisualStudio.InteractiveWindow": "2.0.0-rc3-61304-01",
+    "Microsoft.VisualStudio.Text.Internal": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Language.CallHierarchy": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Language.NavigateTo.Interfaces": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Language.StandardClassification": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Language.Intellisense": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Imaging": "15.0.26014-alpha",
     "Microsoft.VisualStudio.Imaging.Interop.14.0.DesignTime": "14.3.25407",
```
Comment (tannergooding) [inline src/EditorFeatures/Core/project.json]: Shouldn't this be the Dev15 version: https://www.nuget.org/packages/Microsoft.VisualStudio.Imaging.Interop.14.0.DesignTime/15.0.25726-Preview5

File: src/EditorFeatures/Next/project.json [hunk 15]
```diff
@@ -1,22 +1,23 @@
...
-    "Microsoft.VisualStudio.Utilities": "15.0.25726-Preview5",
-    "Microsoft.VisualStudio.Language.NavigateTo.Interfaces": "15.0.25726-Preview5",
-    "Microsoft.VisualStudio.Utilities": "15.0.25726-Preview5",
-    "Microsoft.VisualStudio.Validation": "15.0.55-pre",
-    "Microsoft.VisualStudio.Threading": "15.0.109-pre",
+    "Microsoft.VisualStudio.CoreUtility": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Editor": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Imaging": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Imaging.Interop.14.0.DesignTime": "14.3.25407",
+    "Microsoft.VisualStudio.Shell.Interop.14.0.DesignTime": "14.3.25407",
```
Comment (tannergooding) [inline src/EditorFeatures/Next/project.json]: Same here: https://www.nuget.org/packages/Microsoft.VisualStudio.Shell.Interop.14.0.DesignTime/15.0.25726-Preview5

File: src/EditorFeatures/Next/project.json [hunk 16]
```diff
@@ -1,22 +1,23 @@
...
+    "Microsoft.VisualStudio.Text.UI": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Text.UI.Wpf": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Utilities": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Language.NavigateTo.Interfaces": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Utilities": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Validation": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Threading": "15.0.26014-alpha",
     "System.Collections.Immutable": "1.3.1",
-    "System.Threading.Tasks.Dataflow": "4.5.24",
+    "Microsoft.Tpl.Dataflow": "4.5.24",
```
Comment (tannergooding) [inline src/EditorFeatures/Next/project.json]: Why aren't we consuming the custom SYstem.Threading.Tasks.Dataflow: v4.5.24 package?

File: src/EditorFeatures/Next/project.json [hunk 17]
```diff
@@ -1,22 +1,23 @@
...
     "VS.ExternalAPIs.CodingConventions": "1.0.60704.2"
```
Comment (tannergooding) [inline src/EditorFeatures/Next/project.json]: This shouldn't be needed on Dev15 only. @jasonmalinowski can confirm

File: src/EditorFeatures/Test2/project.json [hunk 18]
```diff
@@ -1,19 +1,21 @@
 ﻿{
   "dependencies": {
-    "System.Threading.Tasks.Dataflow": "4.5.24",
-    "Microsoft.VisualStudio.Composition": "14.2.19-pre",
-    "RoslynDependencies.Microsoft.VisualStudio.Platform.VSEditor": "14.3.25407",
+    "Microsoft.Tpl.Dataflow": "4.5.24",
```
Comment (tannergooding) [inline src/EditorFeatures/Test2/project.json]: Shouldn't this be the custom "System.Threading.Tasks.Dataflow": "4.5.24" package?

File: src/EditorFeatures/Test2/project.json [hunk 19]
```diff
@@ -1,19 +1,21 @@
...
-    "Microsoft.VisualStudio.Language.CallHierarchy": "14.3.25407-alpha",
-    "Microsoft.VisualStudio.Imaging": "14.3.25407",
+    "Microsoft.Build.Runtime": "15.1.0-preview-000458-02",
+    "Microsoft.VisualStudio.InteractiveWindow": "2.0.0-rc3-61304-01",
+    "Microsoft.VisualStudio.Editor": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Language.Intellisense": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Text.Internal": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Language.CallHierarchy": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Imaging": "15.0.26014-alpha",
     "Microsoft.VisualStudio.Imaging.Interop.14.0.DesignTime": "14.3.25407",
```
Comment (tannergooding) [inline src/EditorFeatures/Test2/project.json]: This should probably be the Dev15 version I listed above.

File: src/Tools/Source/SyntaxVisualizer/SyntaxVisualizerExtension/project.json [hunk 20]
```diff
@@ -7,21 +7,23 @@
...
+    "Microsoft.VisualStudio.Shell.15.0": "15.0.26014-alpha",
+    "Microsoft.VisualStudio.Shell.Immutable.10.0": "15.0.26014-alpha",
     "Microsoft.VisualStudio.Shell.Interop": "7.10.6071",
     "Microsoft.VisualStudio.Shell.Interop.8.0": "8.0.50727",
     "Microsoft.VisualStudio.Shell.Interop.9.0": "9.0.30729",
     "Microsoft.VisualStudio.Shell.Interop.10.0": "10.0.30319",
     "Microsoft.VisualStudio.TextManager.Interop": "7.10.6070",
-    "Microsoft.VisualStudio.ComponentModelHost": "14.3.25407-alpha",
+    "Microsoft.VisualStudio.ComponentModelHost": "15.0.26014-alpha",
     "System.Collections.Immutable": "1.1.36",
```
Comment (tannergooding) [inline src/Tools/Source/SyntaxVisualizer/SyntaxVisualizerExtension/project.json]: I think these should be moved forward to the Dev15 versions, if we are moving the VS references forward. @jmarolf may have an opinion here as well.

File: build/config/RepoUtilData.json (hunk 9 above)
Comment (jaredpar) [inline build/config/RepoUtilData.json]: Jenkins still uses it in some cases.  Have to do a round of updates before I can remove this.

File: build/config/RepoUtilData.json (hunk 10 above)
Comment (jaredpar) [inline build/config/RepoUtilData.json]: @tannergooding this is used when building our SDK VSIX package.  It uses our 1.0 drops of Roslyn as the base and we depended on that version of Immutable.

File: build/config/RepoUtilData.json (hunk 11 above)
Comment (jaredpar) [inline build/config/RepoUtilData.json]: Same answer.

File: build/config/RepoUtilData.json (hunk 10 above)
Comment (tannergooding) [inline build/config/RepoUtilData.json]: We can probably do it in a separate PR, but I believe last time I talked to @jmarolf about this, we agreed it was reasonable to move the SDK package forward to v2.0

File: src/Compilers/Extension/project.json (hunk 12 above)
Comment (jaredpar) [inline src/Compilers/Extension/project.json]: Understand the reasoning but disagree based on experience here.  Mixing official and unofficial packages is a nightmare because of the dependency drag.  Until the VS SDK team can publish a *full* and *comprehensive* SDK there is no reasonable way for us to depend on bits and pieces of it.  That is part of the follow up discussion we are having with them in a few weeks.

File: src/Compilers/Extension/project.json (hunk 13 above)
Comment (jaredpar) [inline src/Compilers/Extension/project.json]: Same answer

File: src/Compilers/Extension/project.json (hunk 12 above)
Comment (tannergooding) [inline src/Compilers/Extension/project.json]: I was indicating that we should move to the official RC3 packages where applicable and have the custom packages come from the same build, for maximum compatibility

File: src/EditorFeatures/Next/project.json (hunk 16 above)
Comment (jaredpar) [inline src/EditorFeatures/Next/project.json]: There are official VS packages that we depend on which reference Microsoft.Tpl.Dataflow.  Trying to mix System.Threading.Tasks.Dataflow in with this dependency causes resolve / double write issues in the build.  Hence I had to stop fighting this and move to Microsoft.Tpl.Dataflow. 

I've updated the tracking issue on CoreFX noting the additional pain this is causing.  Once this is available I can contact the VS owners and ask them to adjust their dependencies.  Until then we are stuck. 

Welcome to my nightmare.

Comment (jaredpar) [inline src/EditorFeatures/Next/project.json]: https://github.com/dotnet/corefx/issues/14869

File: src/EditorFeatures/Next/project.json (hunk 17 above)
Comment (jaredpar) [inline src/EditorFeatures/Next/project.json]: I assure you that it is.

File: src/EditorFeatures/Test2/project.json (hunk 18 above)
Comment (jaredpar) [inline src/EditorFeatures/Test2/project.json]: Same answer

File: src/EditorFeatures/Next/project.json (hunk 16 above)
Comment (tannergooding) [inline src/EditorFeatures/Next/project.json]: argh

**Comment by dpoeschl** [comment]
@Pilchie Everything looks fine.

**Comment by Pilchie** [comment]
Great.  Thanks @dpoeschl!

File: src/EditorFeatures/Core/project.json (hunk 14 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/project.json]: @jaredpar Did this matter?

File: src/Tools/Source/SyntaxVisualizer/SyntaxVisualizerExtension/project.json (hunk 20 above)
Comment (jasonmalinowski) [inline src/Tools/Source/SyntaxVisualizer/SyntaxVisualizerExtension/project.json]: @jmarolf Did we want to maintain this running on Dev14?

File: src/EditorFeatures/Core/project.json (hunk 14 above)
Comment (jaredpar) [inline src/EditorFeatures/Core/project.json]: @jasonmalinowski it doesn't affect our ability to run.  It's still unclear what the right thing to do actually is.  It' an interp assembly, target 14.0, whose version is changing.  That is confusing.  I'm going to dig a bit deeper into it and if it needs to change going to submit another issue.

Comment (tannergooding) [inline src/EditorFeatures/Core/project.json]: As long as VS has the appropriate binding redirects it shouldn't, but it would be safer to reference the appropriately versioned package, I would think.

Comment (jaredpar) [inline src/EditorFeatures/Core/project.json]: Part of the confusion is whether or not it's an interop assembly.  The Interop in the name is generally a clue.  If it's interop then the version should never change (unless we have a MS.VS.Shell.Immutable.10.0 situation).
